astropy/uncertainty/core.py

```diff
@@ -43,40 +46,142 @@ def __new__(cls, samples):
+        # regular ndarray, to avoid interference between our complicated dtype and
+        # something else (such as would be the case for Quantity.__array_finalize__,
+        # which needs the dtype to be correct for creating a StructuredUnit).
+        if samples.dtype.itemsize == samples.strides[-1]:
+            # For contiguous last axis, a plain new view suffices.
+            structured = samples.view(np.ndarray)
+        else:
+            # But if the last axis is not contiguous, we use __array_interface__ (as in
+            # np.lib.stridetricks.as_strided) to create a new ndarray where the last
+            # axis is covered by a void of the same size as the structured new_dtype
+            # (assigning it directly does not work: possible empty fields get entries).
+            interface = dict(samples.__array_interface__)
+            interface["descr"] = interface["typestr"] = f"|V{new_dtype.itemsize}"
+            interface["shape"] = samples.shape[:-1]
+            interface["strides"] = samples.strides[:-1]
+            structured = np.asarray(DummyArray(interface, base=samples))
+        # Set our new structured dtype.
+        structured.dtype = new_dtype
+        # Get rid of trailing dimension of 1.
+        structured.shape = samples.shape[:-1]
+
+        # Now view as the Distribution subclass, and finalize based on the
+        # original samples (e.g., to set the unit for QuantityDistribution).
+        new_cls = cls._get_distribution_cls(type(samples))
+        self = structured.view(new_cls)
+        self.__array_finalize__(samples)
+        return self
+
+    @classmethod
+    def _get_distribution_cls(cls, samples_cls):
+        if issubclass(samples_cls, Distribution):
+            return samples_cls
+
         new_cls = cls._generated_subclasses.get(samples_cls)
         if new_cls is None:
-            # Make a new class with the combined name, inserting Distribution
-            # itself below the samples class since that way Quantity methods
-            # like ".to" just work (as .view() gets intercepted).  However,
-            # repr and str are problems, so we put those on top.
-            # TODO: try to deal with this at the lower level.  The problem is
-            # that array2string does not allow one to override how structured
-            # arrays are typeset, leading to all samples to be shown.  It may
-            # be possible to hack oneself out by temporarily becoming a void.
-            new_name = samples_cls.__name__ + cls.__name__
+            # Walk through MRO and find closest base data class.
+            # Note: right now, will basically always be ndarray, but
+            # one could imagine needing some special care for one subclass,
+            # which would then get its own entry.  E.g., if MaskedAngle
+            # defined something special, then MaskedLongitude should depend
+            # on it.
+            for mro_item in samples_cls.__mro__:
```

Comment:
Nitpick. I think 
```suggestion
            for mro_item in samples_cls.mro():
```
is preferred.

Reply:
I didn't know this, but https://stackoverflow.com/questions/19713585/whats-the-difference-between-the-mro-method-and-the-mro-attribute-of-a-clas suggests that `.mro()` is called at class creation time and caches it results in `__mro__`. If so, I happened to have done the right thing...